cpp/src/fil/common.cuh

```diff
@@ -38,79 +38,92 @@ __host__ __device__ __forceinline__ int forest_num_nodes(int num_trees, int dept
   return num_trees * tree_num_nodes(depth);
 }
 
-template <>
-__host__ __device__ __forceinline__ float base_node::output<float>() const
-{
-  return val.f;
-}
-template <>
-__host__ __device__ __forceinline__ int base_node::output<int>() const
-{
-  return val.idx;
-}
-
 /** dense_tree represents a dense tree */
-struct dense_tree {
-  __host__ __device__ dense_tree(dense_node* nodes, int node_pitch)
-    : nodes_(nodes), node_pitch_(node_pitch)
+template <bool can_be_categorical>
```

Comment:
Do we need this parameter in the forest itself? Can't we have the same data structure, but only make the code different?